cosmwasm/enclaves/shared/contract-engine/src/reply_message.rs

```diff
@@ -0,0 +1,320 @@
+        data: decrypted_msg_data,
+    });
+
+    let (id, data_for_validation) = parse_message_id_of_encrypted_reply(input_msg, parsed_reply)?;
+    redact_custom_events(parsed_reply);
+
+    wrap_results_as_parsed_message(input_msg, id, result, parsed_reply, data_for_validation)
+}
+
+fn parse_encrypted_error_reply(
+    input_msg: &SecretMessage,
+    parsed_reply: &mut Reply,
+    response: String,
+) -> Result<ParsedMessage, EnclaveError> {
+    let (id, data_for_validation) = parse_message_id_of_encrypted_reply(input_msg, parsed_reply)?;
+
+    let secret_msg = SecretMessage {
+        nonce: input_msg.nonce,
+        user_public_key: input_msg.user_public_key,
+        msg: base64::decode(response.clone()).map_err(|err| {
+            warn!(
+                "got an error while trying to serialize err reply from base64 {:?}: {}",
+                response, err
+            );
+            EnclaveError::FailedToSerialize
+        })?,
+    };
+
+    let decrypted_error = secret_msg.decrypt()?;
+
+    // Now we need to create synthetic SecretMessage to fit the API in "handle"
+    let result = SubMsgResult::Err(
+        String::from_utf8(decrypted_error[HEX_ENCODED_HASH_SIZE..].to_vec()).map_err(|err| {
+            warn!(
+                "Failed to parse error as string {:?}: {}",
+                decrypted_error[HEX_ENCODED_HASH_SIZE..].to_vec(),
+                err
+            );
+            EnclaveError::FailedToDeserialize
+        })?,
+    );
+
+    wrap_results_as_parsed_message(input_msg, id, result, parsed_reply, data_for_validation)
+}
+
+fn parse_encrypted_reply_message(
+    input_msg: &SecretMessage,
+    parsed_reply: &mut Reply,
+) -> Result<ParsedMessage, EnclaveError> {
+    match parsed_reply.result.clone() {
+        SubMsgResult::Ok(response) => parse_encrypted_ok_reply(input_msg, parsed_reply, response),
+        SubMsgResult::Err(response) => {
+            parse_encrypted_error_reply(input_msg, parsed_reply, response)
+        }
+    }
+}
+
+fn parse_plaintext_reply_message(
+    input_msg: &SecretMessage,
+    parsed_reply: &mut Reply,
+) -> Result<ParsedMessage, EnclaveError> {
+    let msg_id = String::from_utf8(parsed_reply.id.as_slice().to_vec()).map_err(|err| {
+        warn!(
+            "Failed to parse message id as string {:?}: {}",
+            parsed_reply.id.as_slice().to_vec(),
+            err
+        );
+        EnclaveError::FailedToDeserialize
+    })?;
+
+    let msg_id_as_num = match msg_id.parse::<u64>() {
+        Ok(m) => m,
+        Err(err) => {
+            warn!("Failed to parse message id as number {}: {}", msg_id, err);
+            return Err(EnclaveError::FailedToDeserialize);
+        }
+    };
+
+    let decrypted_reply = DecryptedReply {
+        id: msg_id_as_num,
+        result: parsed_reply.result.clone(),
+    };
+
+    redact_custom_events(parsed_reply);
+    let serialized_encrypted_reply: Vec<u8> = serde_json::to_vec(parsed_reply).map_err(|err| {
```

Comment:
probably this should be called serialized_plaintext_reply (since I don't think it's encrypted)? + warning message

Reply:
```suggestion
    let serialized_plaintext_reply: Vec<u8> = serde_json::to_vec(parsed_reply).map_err(|err| {
```